Improve how log is tailed in testclusters on failure

alpar-t opened the pull request:
- only print last few lines
- print all errors and warnings
- compact repeating errors and warnings

Looks like:
```
Standard output of node `node{:x-pack:plugin:ccr:qa:downgrade-to-basic-license:followClusterTest}`
  [2019-03-26T17:06:26,141][WARN ][o.e.n.Node               ] [followClusterTest] version [8.0.0-SNAPSHOT] is a pre-release version of Elasticsearch and is not suitable for production
  [2019-03-26T17:06:30,242][WARN ][o.e.b.BootstrapChecks    ] [followClusterTest] max virtual memory areas vm.max_map_count [65530] is too low, increase to at least [262144]
  [2019-03-26T17:06:30,391][WARN ][o.e.x.s.a.s.m.NativeRoleMappingStore] [followClusterTest] Failed to clear cache for realms [[]]
  [2019-03-26T17:06:33,574][WARN ][o.e.x.c.a.AutoFollowCoordinator] [followClusterTest] failure occurred while fetching cluster state for auto follow pattern [test_pattern]
    ↑ repeated another 5798 times ↑
    ↓ tail of the log ↓
        at io.netty.channel.DefaultChannelPipeline$HeadContext.channelRead(DefaultChannelPipeline.java:1434) [netty-transport-4.1.32.Final.jar:4.1.32.Final]
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:362) [netty-transport-4.1.32.Final.jar:4.1.32.Final]
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:348) [netty-transport-4.1.32.Final.jar:4.1.32.Final]
        at io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:965) [netty-transport-4.1.32.Final.jar:4.1.32.Final]
        at io.netty.channel.nio.AbstractNioByteChannel$NioByteUnsafe.read(AbstractNioByteChannel.java:163) [netty-transport-4.1.32.Final.jar:4.1.32.Final]
        at io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:656) [netty-transport-4.1.32.Final.jar:4.1.32.Final]
        at io.netty.channel.nio.NioEventLoop.processSelectedKeysPlain(NioEventLoop.java:556) [netty-transport-4.1.32.Final.jar:4.1.32.Final]
        at io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:510) [netty-transport-4.1.32.Final.jar:4.1.32.Final]
        at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:470) [netty-transport-4.1.32.Final.jar:4.1.32.Final]
        at io.netty.util.concurrent.SingleThreadEventExecutor$5.run(SingleThreadEventExecutor.java:909) [netty-common-4.1.32.Final.jar:4.1.32.Final]
        at java.lang.Thread.run(Thread.java:835) [?:?]
  [2019-03-26T17:07:36,621][WARN ][o.e.x.c.a.AutoFollowCoordinator] [followClusterTest] failure occurred while fetching cluster state for auto follow pattern [test_pattern]
  org.elasticsearch.ElasticsearchStatusException: can not fetch remote cluster state as the remote cluster [leader_cluster] is not licensed for [ccr]; the license mode [BASIC] on cluster [leader_cluster] does not enable [ccr]

```

alpar-t committed:
Improoce how log is tailed in testclusters on failure

- only print last few lines
- print all errors and warnings
- compact repeating errors and warnings

Looks like:
```
Standard output of node `node{:x-pack:plugin:ccr:qa:downgrade-to-basic-license:followClusterTest}`
  [2019-03-26T17:06:26,141][WARN ][o.e.n.Node               ] [followClusterTest] version [8.0.0-SNAPSHOT] is a pre-release version of Elasticsearch and is not suitable for production
  [2019-03-26T17:06:30,242][WARN ][o.e.b.BootstrapChecks    ] [followClusterTest] max virtual memory areas vm.max_map_count [65530] is too low, increase to at least [262144]
  [2019-03-26T17:06:30,391][WARN ][o.e.x.s.a.s.m.NativeRoleMappingStore] [followClusterTest] Failed to clear cache for realms [[]]
  [2019-03-26T17:06:33,574][WARN ][o.e.x.c.a.AutoFollowCoordinator] [followClusterTest] failure occurred while fetching cluster state for auto follow pattern [test_pattern]
    ↑ repeated another 5798 times ↑
    ↓ tail of the log ↓
        at io.netty.channel.DefaultChannelPipeline$HeadContext.channelRead(DefaultChannelPipeline.java:1434) [netty-transport-4.1.32.Final.jar:4.1.32.Final]
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:362) [netty-transport-4.1.32.Final.jar:4.1.32.Final]
        at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:348) [netty-transport-4.1.32.Final.jar:4.1.32.Final]
        at io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:965) [netty-transport-4.1.32.Final.jar:4.1.32.Final]
        at io.netty.channel.nio.AbstractNioByteChannel$NioByteUnsafe.read(AbstractNioByteChannel.java:163) [netty-transport-4.1.32.Final.jar:4.1.32.Final]
        at io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:656) [netty-transport-4.1.32.Final.jar:4.1.32.Final]
        at io.netty.channel.nio.NioEventLoop.processSelectedKeysPlain(NioEventLoop.java:556) [netty-transport-4.1.32.Final.jar:4.1.32.Final]
        at io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:510) [netty-transport-4.1.32.Final.jar:4.1.32.Final]
        at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:470) [netty-transport-4.1.32.Final.jar:4.1.32.Final]
        at io.netty.util.concurrent.SingleThreadEventExecutor$5.run(SingleThreadEventExecutor.java:909) [netty-common-4.1.32.Final.jar:4.1.32.Final]
        at java.lang.Thread.run(Thread.java:835) [?:?]
  [2019-03-26T17:07:36,621][WARN ][o.e.x.c.a.AutoFollowCoordinator] [followClusterTest] failure occurred while fetching cluster state for auto follow pattern [test_pattern]
  org.elasticsearch.ElasticsearchStatusException: can not fetch remote cluster state as the remote cluster [leader_cluster] is not licensed for [ccr]; the license mode [BASIC] on cluster [leader_cluster] does not enable [ccr]

```

elasticmachine commented:
Pinging @elastic/es-core-infra

rjernst commented on buildSrc/src/main/java/org/elasticsearch/gradle/testclusters/ElasticsearchNode.java:
```diff
@@ -463,15 +467,59 @@ private void logProcessInfo(String prefix, ProcessHandle.Info info) {
+                }
+                String normalizedLine = normalizeLogLine(line);
+                if (errorsAndWarnings.keySet().contains(normalizedLine)) {
+                    logger.lifecycle("  " + line);
```
I think we should use a prefix like we currently do with failures to start in ClusterFormationTasks. It helps visually separate the log contents from normal gradle output.

rjernst commented on buildSrc/src/main/java/org/elasticsearch/gradle/testclusters/ElasticsearchNode.java:
```diff
@@ -463,15 +467,59 @@ private void logProcessInfo(String prefix, ProcessHandle.Info info) {
+            throw new UncheckedIOException(e);
+        }
+        final Map<String, Long> errorsAndWarnings;
+        try(Stream<String> lines = Files.lines(from, StandardCharsets.UTF_8)) {
```
nit: space after try

rjernst commented on buildSrc/src/main/java/org/elasticsearch/gradle/testclusters/ElasticsearchNode.java:
```diff
@@ -463,15 +467,59 @@ private void logProcessInfo(String prefix, ProcessHandle.Info info) {
 
+    private String normalizeLogLine(String line) {
+        if (line.contains("ERROR")) {
+            return line.substring(line.indexOf("ERROR"));
```
Why do you remove ERROR and WARNING prefixes? I think we should keep the log line intact.

atorok replied:
This is just for the normalization part. It's meant as a quick way to remove the timestamp from the start of the line. We never print the normalized line, we print the actual line when we first encounter it and it matches on of the normalized lines.

rjernst replied:
I see the result of this method being placed into the errorsAndWarnings structure, and then later printed from that structure. For example, line 484 and then line 490. What am I missing?

atorok replied:
You're right, I changed it with my last commit to print from there, after I wrote the comment above, but note that we don't remove `ERROR` or `WARN` that's what the line will begin with. We remove anything before them, namely the timestamp that would make each line unique.

rjernst replied:
Timestamps are important. They can help us understand when a particular internal operation is slow.

atorok replied:
@rjernst
We still have the entire log on the file-system for cases like that. 
We would show the time-stamp of the first or last occurrence here anyhow, 
while we complicate the implementation more to retain and show this information I'm not entirely correct that it will be important enough most of the time to do so. 
There will be cases where one needs to download and look at the unmodified log. I would expect that one if which timings are involved will be such a case and the log we show won't matter so much anyhow as it's missing information - such as how frequently the repeated lines are occurring in the logs.

rjernst commented on buildSrc/src/main/java/org/elasticsearch/gradle/testclusters/ElasticsearchNode.java:
```diff
@@ -463,15 +467,59 @@ private void logProcessInfo(String prefix, ProcessHandle.Info info) {
+                if (errorsAndWarnings.keySet().contains(normalizedLine)) {
+                    logger.lifecycle("  " + line);
+                    if (errorsAndWarnings.get(normalizedLine) != 1) {
+                        logger.lifecycle("    ↑ repeated another " + errorsAndWarnings.get(normalizedLine) + " times ↑");
```
Is this normalization accounting for multiline errors, like the same exception being repeated? I think that is important to do.

atorok replied:
I made a change to include exceptions too

rjernst commented on buildSrc/src/main/java/org/elasticsearch/gradle/testclusters/ElasticsearchNode.java:
```diff
@@ -463,15 +467,59 @@ private void logProcessInfo(String prefix, ProcessHandle.Info info) {
+                    }
+                    errorsAndWarnings.remove(normalizedLine);
+                }
+                if (reader.getLineNumber() == lineCutoff) {
```
Why have a cutoff? If there are errors, they are possibly relevant, and more likely more relevant closer to the end of the log (or possibly the first error). I think we should continue to output all errors/warnings.

atorok replied:
We do not cut off errors or warnings. We show the last 50 lines of the log regardless of what it contains and show any error and warning before that.

rjernst reviewed (changes requested):
I like the idea of being a little smarter than what we have in the log excerpting of RestIntegTestTask. But your example output shows a truncated stacktrace from the bottom of the trace. We should keep the entire stack trace. See the "inExcerpt" logic in RestIntegTestTask.

mark-vieira reviewed (commented):
I'll defer to @rjernst's judgement here but I do share some concern around our ability to intelligently deal with stacktraces. Essentially... there be dragons.

I think a lot of the issue with excessive logging is going to be eased by using the Gradle test runner and having the output we want much more easily accessible in build scans. As it is, there is so much other bullshit in our CI build logs, truncating stacktraces seems like it'll have a minimal effect and the risk is we are too overzealous.

rjernst commented:
> truncating stacktraces seems like it'll have a minimal effect and the risk is we are too overzealous

It can be several levels, and I've always found it useful. Stack traces are long enough, no need to see extra stuff that is never relevant. We have done this in Lucene and ES for a very long time; I do think we should continue to clean this up.

mark-vieira commented:
I think something akin to a blacklist (as we had with our test reporter) when we know there are a bunch of internal classes involved that we don't care about is probably safe enough. Trying to detect repeating stacktraces, etc, seems like it might be a bit too much.

alpar-t committed:
PR review

Here's how it looks
```
=== Standard output of node `node{:x-pack:plugin:ccr:qa:downgrade-to-basic-license:followClusterTest}` ===

»    ↓ errors and warnings from /home/alpar/work/elastic/elasticsearch/x-pack/plugin/ccr/qa/downgrade-to-basic-license/build/testclusters/followClusterTest/logs/es.stdout.log ↓
» WARN ][o.e.x.s.a.s.m.NativeRoleMappingStore] [followClusterTest] Failed to clear cache for realms [[]]
» WARN ][o.e.x.c.a.AutoFollowCoordinator] [followClusterTest] failure occurred while fetching cluster state for auto follow pattern [test_pattern]
»  org.elasticsearch.ElasticsearchStatusException: can not fetch remote cluster state as the remote cluster [leader_cluster] is not licensed for [ccr]; the license mode [BASIC] on cluster [leader_cluster] does not enable [ccr]
»       at org.elasticsearch.xpack.ccr.CcrLicenseChecker.clusterStateNonCompliantRemoteLicense(CcrLicenseChecker.java:416) ~[x-pack-ccr-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
»       at org.elasticsearch.xpack.ccr.CcrLicenseChecker$1.onResponse(CcrLicenseChecker.java:220) [x-pack-ccr-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
»       at org.elasticsearch.xpack.ccr.CcrLicenseChecker$1.onResponse(CcrLicenseChecker.java:210) [x-pack-ccr-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
»       at org.elasticsearch.license.RemoteClusterLicenseChecker$1.onResponse(RemoteClusterLicenseChecker.java:172) [x-pack-core-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
»       at org.elasticsearch.license.RemoteClusterLicenseChecker$1.onResponse(RemoteClusterLicenseChecker.java:161) [x-pack-core-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
»       at org.elasticsearch.action.support.ContextPreservingActionListener.onResponse(ContextPreservingActionListener.java:43) [elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
»       at org.elasticsearch.action.ActionListenerResponseHandler.handleResponse(ActionListenerResponseHandler.java:54) [elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
»       at org.elasticsearch.transport.TransportService$ContextRestoreResponseHandler.handleResponse(TransportService.java:1095) [elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
»       at org.elasticsearch.transport.InboundHandler$1.doRun(InboundHandler.java:224) [elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
»       at org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:37) [elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
»       at org.elasticsearch.common.util.concurrent.EsExecutors$DirectExecutorService.execute(EsExecutors.java:192) [elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
»       at org.elasticsearch.transport.InboundHandler.handleResponse(InboundHandler.java:216) [elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
»       at org.elasticsearch.transport.InboundHandler.messageReceived(InboundHandler.java:141) [elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
»       at org.elasticsearch.transport.InboundHandler.inboundMessage(InboundHandler.java:105) [elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
»       at org.elasticsearch.transport.TcpTransport.inboundMessage(TcpTransport.java:644) [elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
»       at org.elasticsearch.transport.netty4.Netty4MessageChannelHandler.channelRead(Netty4MessageChannelHandler.java:62) [transport-netty4-client-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
»       at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:362) [netty-transport-4.1.32.Final.jar:4.1.32.Final]
»       at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:348) [netty-transport-4.1.32.Final.jar:4.1.32.Final]
»       at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:340) [netty-transport-4.1.32.Final.jar:4.1.32.Final]
»       at io.netty.handler.codec.ByteToMessageDecoder.fireChannelRead(ByteToMessageDecoder.java:323) [netty-codec-4.1.32.Final.jar:4.1.32.Final]
»       at io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:297) [netty-codec-4.1.32.Final.jar:4.1.32.Final]
»       at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:362) [netty-transport-4.1.32.Final.jar:4.1.32.Final]
»       at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:348) [netty-transport-4.1.32.Final.jar:4.1.32.Final]
»       at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:340) [netty-transport-4.1.32.Final.jar:4.1.32.Final]
»       at io.netty.handler.logging.LoggingHandler.channelRead(LoggingHandler.java:241) [netty-handler-4.1.32.Final.jar:4.1.32.Final]
»       at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:362) [netty-transport-4.1.32.Final.jar:4.1.32.Final]
»       at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:348) [netty-transport-4.1.32.Final.jar:4.1.32.Final]
»       at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:340) [netty-transport-4.1.32.Final.jar:4.1.32.Final]
»       at io.netty.channel.DefaultChannelPipeline$HeadContext.channelRead(DefaultChannelPipeline.java:1434) [netty-transport-4.1.32.Final.jar:4.1.32.Final]
»       at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:362) [netty-transport-4.1.32.Final.jar:4.1.32.Final]
»       at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:348) [netty-transport-4.1.32.Final.jar:4.1.32.Final]
»       at io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:965) [netty-transport-4.1.32.Final.jar:4.1.32.Final]
»       at io.netty.channel.nio.AbstractNioByteChannel$NioByteUnsafe.read(AbstractNioByteChannel.java:163) [netty-transport-4.1.32.Final.jar:4.1.32.Final]
»       at io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:656) [netty-transport-4.1.32.Final.jar:4.1.32.Final]
»       at io.netty.channel.nio.NioEventLoop.processSelectedKeysPlain(NioEventLoop.java:556) [netty-transport-4.1.32.Final.jar:4.1.32.Final]
»       at io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:510) [netty-transport-4.1.32.Final.jar:4.1.32.Final]
»       at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:470) [netty-transport-4.1.32.Final.jar:4.1.32.Final]
»       at io.netty.util.concurrent.SingleThreadEventExecutor$5.run(SingleThreadEventExecutor.java:909) [netty-common-4.1.32.Final.jar:4.1.32.Final]
»       at java.lang.Thread.run(Thread.java:835) [?:?]
»   ↑ repeated 5681 times ↑
»   ↓ last 40 non error or warning messages from /home/alpar/work/elastic/elasticsearch/x-pack/plugin/ccr/qa/downgrade-to-basic-license/build/testclusters/followClusterTest/logs/es.stdout.log ↓
» [2019-03-29T19:02:40,012][INFO ][o.e.x.m.p.NativeController] [followClusterTest] Native controller process has stopped - no new native processes can be started

=== Standard output of node `node{:x-pack:plugin:ccr:qa:downgrade-to-basic-license:leaderClusterTest}` ===

»    ↓ errors and warnings from /home/alpar/work/elastic/elasticsearch/x-pack/plugin/ccr/qa/downgrade-to-basic-license/build/testclusters/leaderClusterTest/logs/es.stdout.log ↓
» WARN ][o.e.x.s.a.s.m.NativeRoleMappingStore] [leaderClusterTest] Failed to clear cache for realms [[]]
»   ↓ last 40 non error or warning messages from /home/alpar/work/elastic/elasticsearch/x-pack/plugin/ccr/qa/downgrade-to-basic-license/build/testclusters/leaderClusterTest/logs/es.stdout.log ↓
» [2019-03-29T19:01:22,419][INFO ][o.e.p.PluginsService     ] [leaderClusterTest] loaded module [x-pack-ilm]

```

atorok commented:
I implemented the suggestions  except changing the stack trace of the exceptions. 
we didn't have this here before and we can always add it in a follow-up 
This is how the log now looks:
```
=== Standard output of node `node{:x-pack:plugin:ccr:qa:downgrade-to-basic-license:followClusterTest}` ===

»    ↓ errors and warnings from /home/alpar/work/elastic/elasticsearch/x-pack/plugin/ccr/qa/downgrade-to-basic-license/build/testclusters/followClusterTest/logs/es.stdout.log ↓
» WARN ][o.e.x.s.a.s.m.NativeRoleMappingStore] [followClusterTest] Failed to clear cache for realms [[]]
» WARN ][o.e.x.c.a.AutoFollowCoordinator] [followClusterTest] failure occurred while fetching cluster state for auto follow pattern [test_pattern]
»  org.elasticsearch.ElasticsearchStatusException: can not fetch remote cluster state as the remote cluster [leader_cluster] is not licensed for [ccr]; the license mode [BASIC] on cluster [leader_cluster] does not enable [ccr]
»       at org.elasticsearch.xpack.ccr.CcrLicenseChecker.clusterStateNonCompliantRemoteLicense(CcrLicenseChecker.java:416) ~[x-pack-ccr-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
»       at org.elasticsearch.xpack.ccr.CcrLicenseChecker$1.onResponse(CcrLicenseChecker.java:220) [x-pack-ccr-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
»       at org.elasticsearch.xpack.ccr.CcrLicenseChecker$1.onResponse(CcrLicenseChecker.java:210) [x-pack-ccr-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
»       at org.elasticsearch.license.RemoteClusterLicenseChecker$1.onResponse(RemoteClusterLicenseChecker.java:172) [x-pack-core-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
»       at org.elasticsearch.license.RemoteClusterLicenseChecker$1.onResponse(RemoteClusterLicenseChecker.java:161) [x-pack-core-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
»       at org.elasticsearch.action.support.ContextPreservingActionListener.onResponse(ContextPreservingActionListener.java:43) [elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
»       at org.elasticsearch.action.ActionListenerResponseHandler.handleResponse(ActionListenerResponseHandler.java:54) [elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
»       at org.elasticsearch.transport.TransportService$ContextRestoreResponseHandler.handleResponse(TransportService.java:1095) [elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
»       at org.elasticsearch.transport.InboundHandler$1.doRun(InboundHandler.java:224) [elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
»       at org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:37) [elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
»       at org.elasticsearch.common.util.concurrent.EsExecutors$DirectExecutorService.execute(EsExecutors.java:192) [elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
»       at org.elasticsearch.transport.InboundHandler.handleResponse(InboundHandler.java:216) [elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
»       at org.elasticsearch.transport.InboundHandler.messageReceived(InboundHandler.java:141) [elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
»       at org.elasticsearch.transport.InboundHandler.inboundMessage(InboundHandler.java:105) [elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
»       at org.elasticsearch.transport.TcpTransport.inboundMessage(TcpTransport.java:644) [elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
»       at org.elasticsearch.transport.netty4.Netty4MessageChannelHandler.channelRead(Netty4MessageChannelHandler.java:62) [transport-netty4-client-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]
»       at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:362) [netty-transport-4.1.32.Final.jar:4.1.32.Final]
»       at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:348) [netty-transport-4.1.32.Final.jar:4.1.32.Final]
»       at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:340) [netty-transport-4.1.32.Final.jar:4.1.32.Final]
»       at io.netty.handler.codec.ByteToMessageDecoder.fireChannelRead(ByteToMessageDecoder.java:323) [netty-codec-4.1.32.Final.jar:4.1.32.Final]
»       at io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:297) [netty-codec-4.1.32.Final.jar:4.1.32.Final]
»       at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:362) [netty-transport-4.1.32.Final.jar:4.1.32.Final]
»       at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:348) [netty-transport-4.1.32.Final.jar:4.1.32.Final]
»       at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:340) [netty-transport-4.1.32.Final.jar:4.1.32.Final]
»       at io.netty.handler.logging.LoggingHandler.channelRead(LoggingHandler.java:241) [netty-handler-4.1.32.Final.jar:4.1.32.Final]
»       at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:362) [netty-transport-4.1.32.Final.jar:4.1.32.Final]
»       at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:348) [netty-transport-4.1.32.Final.jar:4.1.32.Final]
»       at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:340) [netty-transport-4.1.32.Final.jar:4.1.32.Final]
»       at io.netty.channel.DefaultChannelPipeline$HeadContext.channelRead(DefaultChannelPipeline.java:1434) [netty-transport-4.1.32.Final.jar:4.1.32.Final]
»       at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:362) [netty-transport-4.1.32.Final.jar:4.1.32.Final]
»       at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:348) [netty-transport-4.1.32.Final.jar:4.1.32.Final]
»       at io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:965) [netty-transport-4.1.32.Final.jar:4.1.32.Final]
»       at io.netty.channel.nio.AbstractNioByteChannel$NioByteUnsafe.read(AbstractNioByteChannel.java:163) [netty-transport-4.1.32.Final.jar:4.1.32.Final]
»       at io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:656) [netty-transport-4.1.32.Final.jar:4.1.32.Final]
»       at io.netty.channel.nio.NioEventLoop.processSelectedKeysPlain(NioEventLoop.java:556) [netty-transport-4.1.32.Final.jar:4.1.32.Final]
»       at io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:510) [netty-transport-4.1.32.Final.jar:4.1.32.Final]
»       at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:470) [netty-transport-4.1.32.Final.jar:4.1.32.Final]
»       at io.netty.util.concurrent.SingleThreadEventExecutor$5.run(SingleThreadEventExecutor.java:909) [netty-common-4.1.32.Final.jar:4.1.32.Final]
»       at java.lang.Thread.run(Thread.java:835) [?:?]
»   ↑ repeated 5681 times ↑
»   ↓ last 40 non error or warning messages from /home/alpar/work/elastic/elasticsearch/x-pack/plugin/ccr/qa/downgrade-to-basic-license/build/testclusters/followClusterTest/logs/es.stdout.log ↓
» [2019-03-29T19:02:40,012][INFO ][o.e.x.m.p.NativeController] [followClusterTest] Native controller process has stopped - no new native processes can be started

=== Standard output of node `node{:x-pack:plugin:ccr:qa:downgrade-to-basic-license:leaderClusterTest}` ===

»    ↓ errors and warnings from /home/alpar/work/elastic/elasticsearch/x-pack/plugin/ccr/qa/downgrade-to-basic-license/build/testclusters/leaderClusterTest/logs/es.stdout.log ↓
» WARN ][o.e.x.s.a.s.m.NativeRoleMappingStore] [leaderClusterTest] Failed to clear cache for realms [[]]
»   ↓ last 40 non error or warning messages from /home/alpar/work/elastic/elasticsearch/x-pack/plugin/ccr/qa/downgrade-to-basic-license/build/testclusters/leaderClusterTest/logs/es.stdout.log ↓
» [2019-03-29T19:01:22,419][INFO ][o.e.p.PluginsService     ] [leaderClusterTest] loaded module [x-pack-ilm]
...
```

So the multi line log messages are combined back together, we can filter out some annoyances that will always be there and we don't care about and also avoid repeating the errors and warnings when we tail the output. 
Exceptions will always be shown in full or not at all due to the messages being combined back together, and us cutting off across message boundaries.

alpar-t committed:
remove unused import

alpar-t committed:
Merge remote-tracking branch 'origin/master' into testclusters-cluster-log-improovements

alpar-t committed:
Merge remote-tracking branch 'origin/master' into testclusters-cluster-log-improovements

alpar-t committed:
Fix distribution group  config

alpar-t committed:
Fix test

atorok commented:
@rjernst @mark-vieira ready for a new review

mark-vieira commented on buildSrc/src/main/java/org/elasticsearch/gradle/testclusters/TestClustersPlugin.java:
```diff
@@ -327,6 +322,7 @@ private static void autoConfigureClusterDependencies(
                 Map<String, Object> projectNotation = new HashMap<>();
                 projectNotation.put("path", unreleasedInfo.gradleProjectPath);
                 projectNotation.put("configuration", esNode.getDistribution().getLiveConfiguration());
+                logger.debug("Cluster {} depends on {}", esNode.getName(), projectNotation);
```
Debug logging is so noisy it's basically useless. If I'm _actually_ debugging I'm just going to attach a debugger to the build. If we are going to bury these messages under debug logging we should consider whether they are even valuable.

mark-vieira commented on buildSrc/src/main/java/org/elasticsearch/gradle/testclusters/ElasticsearchNode.java:
```diff
@@ -466,48 +472,67 @@ private void logProcessInfo(String prefix, ProcessHandle.Info info) {
-        }
-        lineCutoff -= 50;
+        final Map<String, Integer> errorsAndWarnings = new LinkedHashMap<>();
+        LinkedList<String> ring = new LinkedList<>();
```
I'm not sure I understand the name "ring" here.

rjernst replied:
It's a [circular buffer](https://en.wikipedia.org/wiki/Circular_buffer), although we are using it more like a deque without its second level indirection overhead.

mark-vieira commented on buildSrc/src/main/java/org/elasticsearch/gradle/testclusters/ElasticsearchNode.java:
```diff
@@ -462,14 +472,77 @@ private void logProcessInfo(String prefix, ProcessHandle.Info info) {
+                    // warning as we want to show all of them
+                    if (ring.isEmpty() == false) {
+                        String previousMessage = normalizeLogLine(ring.getLast());
+                        if (MESSAGES_WE_DONT_CARE_ABOUT.stream().anyMatch(match -> previousMessage.contains(match))) {
```
How could this ever be the case if we check for this condition before adding to the collection on line 505?

atorok replied:
good catch!  It was meant to allow for ignoring stack traces with their message too, but it wasn't working as intended. 
Removed now.

mark-vieira commented on buildSrc/src/main/java/org/elasticsearch/gradle/testclusters/ElasticsearchNode.java:
```diff
@@ -462,14 +472,77 @@ private void logProcessInfo(String prefix, ProcessHandle.Info info) {
+                        lineToAdd = ring.removeLast() + "\n" + line;
+                    }
+                }
+                if (MESSAGES_WE_DONT_CARE_ABOUT.stream().anyMatch(match -> lineToAdd.contains(match)) == false) {
```
Couldn't we check the raw line for this first thing then we could shirt-circuit all the other stuff?

atorok replied:
I updated this logic to filter at the end. Raw line wouldn't work for stack-traces.

mark-vieira commented on buildSrc/src/testKit/testclusters/build.gradle:
```diff
@@ -63,10 +63,11 @@ task printLog {
-task upToDate2 {
-    useCluster testClusters.myTestCluster
+    outputs.upToDateWhen { true }
+    inputs.file(file("build.gradle"))
```
I think hard-coding `upToDateWhen` makes declaring this input unnecessary.

atorok replied:
I think it was needed since the task has no inputs.

mark-vieira replied:
I tested this locally and it works fine with a task with no inputs.

mark-vieira reviewed (approved):
Minor code items. Otherwise I'm deferring the functional 👍 here to @rjernst.

rjernst commented on buildSrc/src/main/java/org/elasticsearch/gradle/testclusters/ElasticsearchNode.java:
```diff
@@ -466,48 +472,67 @@ private void logProcessInfo(String prefix, ProcessHandle.Info info) {
+                    if (ring.isEmpty()) {
+                        lineToAdd = line;
+                    } else {
+                        lineToAdd = ring.removeLast() + "\n" + line;
```
This is going to be very inefficient, creating tons of strings whenever we encounter an exception. I wonder if we really need the de-duplication (as @mark-vieira suggested earlier). The algorithm we have now in RestIntegTestTask doesn't require any additional memory. It does mean we lean towards the beginning of the log instead of the end, but I think this is ok.

atorok replied:
@rjernst I do think the de-duplication is useful, I got inspired to implement it by looking at test failures when converting stuff to test clusters, and otherwise I think we tend  to have repeating warnings in the logs when something happens. Makes it easy to tell if something else was also going on that would be hard to fin between the repeating lines. 
I'm not sure it the inefficiency will be noticeable for anything but huge logs with lots of exceptions, if we're concerned abut that we could tail  e.x. the last 10K of the log

rjernst replied:
> I'm not sure it the inefficiency will be noticeable for anything but huge logs with lots of exceptions

This is exactly the case when we don't the build to fail on us. I don't think we should add anything blatantly unperformant when large logs occur since they do happen.

I think we should keep the original single pass that we have in RestIntegTestTask. If we want to have deduplication, it should be best effort, using a cache of the last N messages seen.

alpar-t committed:
Merge remote-tracking branch 'origin/master' into testclusters-cluster-log-improovements

alpar-t committed:
PR review

atorok commented:
@mark-vieira @rjernst  ready for another review.

mark-vieira reviewed (approved):
👍

alpar-t committed:
Remove unnecesary tasks inpu

alpar-t committed:
Merge remote-tracking branch 'origin/master' into testclusters-cluster-log-improovements

alpar-t committed:
Merge remote-tracking branch 'origin/master' into testclusters-cluster-log-improovements

atorok commented:
@rjernst  does this lgty now ?

alpar-t committed:
Merge remote-tracking branch 'origin/master' into testclusters-cluster-log-improovements
